File: solutions/99/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Nov 29 16:02:06 2021
> Description:   
 ************************************************************************/
#include <memory> 
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

template <typename T> 
struct TreeNode {
    T val; 
    shared_ptr<TreeNode<T>> left; 
    shared_ptr<TreeNode<T>> right; 
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(T x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(T x, shared_ptr<TreeNode<T>> left, shared_ptr<TreeNode<T>> right) :
        val(x), left(left), right(right) {}
};

template <typename T> 
class Solution {
private:
    shared_ptr<TreeNode<T>> pre;
    shared_ptr<TreeNode<T>> node1;
    shared_ptr<TreeNode<T>> node2;
public:
    Solution() : pre(nullptr), node1(nullptr), node2(nullptr) {}
    void traverse(shared_ptr<TreeNode<T>> node) {
        if (node == nullptr) return;
        // in order traversal
        traverse(node->left);
        if (pre->val > node->val) {
            if (node1 == nullptr) {
                node1 = pre;
            }
            else {
                node2 = node;
            }
        }
        pre = node;
        traverse(node->right);
    }
    void recoverTree(shared_ptr<TreeNode<T>> root) {
        pre = make_shared<TreeNode<T>>(INT_MIN);
        traverse(root);
        if (node2 == nullptr) {
            pre = make_shared<TreeNode<T>>(INT_MIN);
            traverse(root);
        }
        swap(node1->val, node2->val);
    }
};

TEST(Test99, SimpleTest) {
    auto root = make_shared<TreeNode<int>>(1);
    auto left = make_shared<TreeNode<int>>(3);
    auto right = make_shared<TreeNode<int>>(2);
    root->left = left; 
    left->right = right;
    Solution<int> s; 
    s.recoverTree(root);
    EXPECT_EQ(root->val, 3);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/99/main.py
from typing import Optional 

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self): 
        self.node1 = None 
        self.node2 = None 
        self.pre = None 
    def traverse(self, node): 
        if node is None: 
            return 
        # left 
        self.traverse(node.left)
        # visit 
        if node.val < self.pre.val: 
            if self.node1 is None: 
                self.node1 = self.pre  
            elif self.node2 is None: 
                self.node2 = node 
        self.pre = node 
        # right 
        self.traverse(node.right)
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.pre = TreeNode(-float('inf'))  
        self.traverse(root)
        if self.node2 is None: 
            self.pre = TreeNode(-float('inf')) 
            self.traverse(root)
        temp = self.node1.val 
        self.node1.val = self.node2.val 
        self.node2.val = temp 

if __name__ == "__main__": 
    root = TreeNode(5)
    left = TreeNode(3)
    right = TreeNode(9)
    root.left = left 
    root.right = right 
    left1 = TreeNode(-2147483648)
    right1 = TreeNode(2)
    left.left = left1 
    left.right = right1
    s = Solution()
    s.recoverTree(root)
    assert root.left.val == 2
